wagtail/wagtailadmin/utils.py

```diff
@@ -52,16 +52,30 @@ def get_object_usage(obj):
 
 def users_with_page_permission(page, permission_type, include_superusers=True):
     # Get user model
+    if not include_superusers:
+        raise NotImplementedError("This functionality is no longer supported")
+
     User = get_user_model()
 
-    # Find GroupPagePermission records of the given type that apply to this page or an ancestor
-    ancestors_and_self = list(page.get_ancestors()) + [page]
-    perm = GroupPagePermission.objects.filter(permission_type=permission_type, page__in=ancestors_and_self)
-    q = Q(groups__page_permissions=perm)
+    # Get all page permissions for all users (this is going to be slow if you
+    # have lots of users). Note that the `permissions` attribute isn't set on
+    # `PagePermissionTester` if the user is a superuser or not active, so we
+    # pre-filter out such users to avoid attribtue errors. We OR superusers
+    # back in later, and we don't care about inactive users.
+    user_page_permissions = [
+        page.permissions_for_user(user)
+        for user in User.objects.filter(is_superuser=False, is_active=True)
```

Comment:
I don't think that's a good idea - the only thing that distinguishes front-end and back-end users is the 'can access Wagtail admin' permission, and I'd rather not have anything depend on that at the model level. (I can easily imagine us dealing with editor permissions that don't go through the admin in the near future - through the API, say.)

Overall I'm not really keen on the approach taken by this PR - I don't think it's possible to efficiently perform forward ("what permission does user X have") and reverse ("which users have permission Y") permission checks with the same logic. I think if we're going for the 'permission module' approach as per #1384, we should have both the forward and reverse logic in that one module - and a developer who wants to customise the permission logic would implement both of those methods.